enforcer/patch_manager.py

```diff
@@ -0,0 +1,254 @@
+    except ValueError:
+        logger.warning(f"Invalid CPU string format: {cpu_str}")
+        return None
+
+
+def to_mem_bytes(mem_str: Optional[str]) -> Optional[int]:
+    """
+    Convert Kubernetes memory request string to bytes.
+
+    Args:
+        mem_str: Memory string like "128Mi", "1Gi", "256M", "1000000000", or None
+
+    Returns:
+        Integer number of bytes or None if input is None
+
+    Examples:
+        "128Mi" -> 134217728 (128 * 1024 * 1024)
+        "1Gi" -> 1073741824 (1 * 1024 * 1024 * 1024)
+        "256M" -> 256000000 (256 * 1000 * 1000)
+        "1000" -> 1000
+        None -> None
+    """
+    if mem_str is None:
+        return None
+
+    mem_str = mem_str.strip()
+    if not mem_str:
+        return None
+
+    # Binary (base 1024) suffixes
+    binary_suffixes = {
+        'Ki': 1024,
+        'Mi': 1024 ** 2,
+        'Gi': 1024 ** 3,
+        'Ti': 1024 ** 4,
+        'Pi': 1024 ** 5,
+        'Ei': 1024 ** 6,
+    }
+
+    # Decimal (base 1000) suffixes
+    decimal_suffixes = {
+        'k': 1000,
+        'M': 1000 ** 2,
+        'G': 1000 ** 3,
+        'T': 1000 ** 4,
+        'P': 1000 ** 5,
+        'E': 1000 ** 6,
+    }
+
+    # Check binary suffixes first (more common in K8s)
+    for suffix, multiplier in binary_suffixes.items():
+        if mem_str.endswith(suffix):
+            try:
+                return int(float(mem_str[:-len(suffix)]) * multiplier)
+            except ValueError:
+                logger.warning(f"Invalid memory string format: {mem_str}")
+                return None
+
+    # Check decimal suffixes
+    for suffix, multiplier in decimal_suffixes.items():
+        if mem_str.endswith(suffix):
+            try:
+                return int(float(mem_str[:-len(suffix)]) * multiplier)
+            except ValueError:
+                logger.warning(f"Invalid memory string format: {mem_str}")
+                return None
+
+    # No suffix, assume bytes
+    try:
+        return int(float(mem_str))
+    except ValueError:
+        logger.warning(f"Invalid memory string format: {mem_str}")
+        return None
+
+
+def significant_diff(old: Optional[float], new: float, percent_threshold: float) -> bool:
+    if not old:  # old cpu is none or 0
+        return True
+
+    percent_diff = abs(new - old) / abs(old) * 100
+    return percent_diff > percent_threshold
+
+
+def add_resource_value(resources: Dict[str, Any], resource_type: str, resource_name: str, resource_value: Any) -> None:
+    if resource_type not in resources:
+        resources[resource_type] = {}
+    resources[resource_type][resource_name] = str(resource_value)
+
+
+def get_updated_resources(resources: Dict[str, Any], recommendation: ResourcesRecommendation) -> Dict[str, Any]:
+    if recommendation.cpu:
+        old_cpu_req = to_cpu_num(resources.get(REQ, {}).get(CPU))
+        if old_cpu_req:
+            if significant_diff(old_cpu_req, recommendation.cpu.request, UPDATE_THRESHOLD):
```

Comment:
Consider making a single decision about significant_diff as opposed to deciding separately for the request and limit.

The current behaviour can lead to a scenario where validate_resources fails. E.g. if:

request=1.1
recommended_request=1.0
limit=2
recommended_limit=1.0

In this case you will leave request unmodified at 1.1 (because the diff is not significant) yet you will attempt (and fail) to modify the limit to a value beneath the request.